design-documents/service-isolation/detailed-architecture.md

```diff
@@ -0,0 +1,87 @@
+
+Auth JWT, Customer Context JWT and Store Context are all optional, however, when present in the original client request, they MUST be retransmitted along with all subsequent calls to Magento services. They should not be retransmitted with the calls to the third-party systems.
+ 
+The following sequence diagram demonstrates the process of client application obtaining authentication and customer context JWTs.
+
+![Customer authentication](img/BFF%20-%20Customer%20Authentication%20Sequence.png)
+ 
+## Query enrichment
+
+When the user visits Magento storefront for the first time, pages that support scoping (like category page) must be rendered based on default scopes.
+Client application may send query with no currency, store code and customer context. At the same time, catalog storefront API requires store code, currency and customer group to be passed as explicit arguments. 
+The solution is to have GraphQL query enriched by the GraphQL server. Enrichment logic cannot be moved to storefront gateway because of couple reasons:
+ - it requires GraphQL query to be parsed
+ - it relies on customer context arguments to be decoded. This can be done only by GraphQL server which owns corresponding encoding secret
+ 
+Data lookup for storefront API query enrichment will be done using the following sources in the specified order:
+ 1. Customer ID will always be taken from the Authentication JWT
+ 1. Customer group, customer segment and any extensions to the customer context will always be taken from the Customer Context JWT
+ 1. Data from GraphQL query body
+ 1. Data from store context headers
+ 1. Defaults from GraphQL Server cache
+ 1. Preconfigured defaults via calls to storefront services. For example, default customer group based on current store code. The list of defaults with corresponding sources MUST be configurable
+ 
+The following diagram explains a sequence of actions needed to render a product page with PWA. After the initial call caches on server, gateway and client side will make this sequence shorter.
+
+![Product details page](img/PWA%20BFF%20-%20Product%20Page%20Sequence.png)
+
+## Scalability
+
+All components must support scalability.
+
+## Extensibility
+
+Extensibility is extremely important for Magento and multi-service Magento is not an exception. 
+
+Most of the components need to be extensible, including:
+ - GraphQL Server and any custom BFFs
+ - PWA
+ - Domain services
+ - List of supported authentication mechanisms
+ - Access control list (ACL) 
+ 
+Storefront Gateway may not need to be extensible.
+
+It is important to note that services have to be redeployed after installation of extensions. It should be possible to manage ACL via admin API gateway to avoid the need of redeployment.
+
+## Open questions
```

Comment:
Should we also support replaceability? We had [this](https://github.com/magento/architecture/pull/96) proposal some time ago.

Reply:
Replaceability should be supported based on the service API contract. Custom implementation could be provided as soon as they comply with the API.

The referenced proposal was closed as far as I can see.